Use 'mkdirp' to create log directory if necessary

thebinarypenguin opened the pull request.

thebinarypenguin committed:
Use 'mkdirp' to create log directory if necessary

thebinarypenguin commented:
I just realized that my test doesn't cleanup after itself. It creates the directories `test/foo/bar/` and doesn't remove them. This also raises the topic of added dependencies. I could add `rimraf` and use it to cleanup the test. Or since I'm already including `mkdirp` maybe I should just add the all-encompassing `fs-extra`. 

What are your thoughts on this?

DavidTPate commented on lib/utils.js:
```diff
@@ -13,16 +15,19 @@ const internals = {};
-        stack: err.stack,
-        timestamp: Date.now()
-    };
+    Mkdirp(Path.dirname(options.logPath), () => {
```
Why aren't we handling the error that could be returned from `mkdirp`? I would expect that if we have an error with `mkdirp` that we would propagate that to the callback.

thebinarypenguin replied:
Well a few things...

1. The calling function doesn't make use of it.

2. It seems a bit inconsequential seeing as how if we've gotten this far then we've an uncaught exception and the process is about to exit.

3. Adding the error handling code to the log function is obviously easy, but what do you think the plugin's calling function should do with the error? 

4. Also I'm not sure how to create a good cross-platform test for the err handling code. Any thoughts on that?

I'll add the error handling code in a few minutes but it'll take me longer to devise a test.

DavidTPate replied:
It seems like the best method to cover that scenario would be to mock the `Mkdirp` call and just make it always return an error for that test.

thebinarypenguin replied:
I've never used mocks before so bear with me. It sounds like you are telling me to 

1. create a new "fake" `Mkdirp` function that does nothing but call the callback with an Error
2. **Somehow** inject the reference to the new fake function into `PoopUtils.log` thereby replacing the "real" `Mkdirp` function
3. call `PoopUtils.log` and assert that the callback is called with an error (which confirms that the "fake" function was called and not the "real" function.

What I don't understand is how to inject the "fake" `Mkdirp` into `PoopUtils.log`. Help.

thebinarypenguin replied:
Nevermind, I discovered [proxyquire](https://github.com/thlorenz/proxyquire) that'll do it

thebinarypenguin committed:
Handle error that may result from mkdirp

thebinarypenguin committed:
Make test cleanup after itself

The "creates the log directory if necessary" test now deletes the directories
it creates

thebinarypenguin committed:
Add missing return before callback()

thebinarypenguin committed:
Use proxyquire to mock a mkdirp error

thebinarypenguin commented:
What's the status on this? I think I've fixed all the mentioned issues.

hueniverse commented:
@cjihrig What's the story with this?
